File: src/test/java/io/javalin/ssl/plugin/SSLConfigTest.java
package io.javalin.ssl.plugin;

import io.javalin.ssl.plugin.SSLConfig;
import io.javalin.ssl.plugin.SSLConfigException;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.io.InputStream;
import java.nio.file.Path;
import java.nio.file.Paths;

import static org.junit.jupiter.api.Assertions.*;

@Tag("unitary")
class SSLConfigTest {

    final String absolutePathString = "/etc/sample/path";
    final Path absolutePath = Paths.get(absolutePathString);

    final String fileName = "sample.pem";

    final InputStream inputStream = InputStream.nullInputStream();

    final String pemString =
            "----- START CERTIFICATE -----\n" +
            "BLABLABLABLABLABLABLABLABLABLA" +
            "----- END CERTIFICATE -----\n";


    //////////////////////////////
    // Pem loading tests        //
    //////////////////////////////

    @Test
    void loadPemFromPathCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromPath(absolutePathString,absolutePathString));
        assertEquals(absolutePath, config.inner.pemCertificatesPath);
        assertEquals(absolutePath, config.inner.pemPrivateKeyPath);
    }

    @Test
    void loadPemFromPathDifferentMethod() {
        SSLConfig config = new SSLConfig();
        config.pemFromString("", ""); // load empty strings
        assertThrows(SSLConfigException.class,() -> config.pemFromPath(absolutePathString,absolutePathString));
    }

    @Test
    void loadPemFromPathTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromPath(absolutePathString,absolutePathString);
        assertThrows(SSLConfigException.class,() -> config.pemFromPath(absolutePathString,absolutePathString));
    }

    @Test
    void loadPemFromPathWithPasswordCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromPath(absolutePathString,absolutePathString, "password"));
        assertEquals(absolutePath, config.inner.pemCertificatesPath);
        assertEquals(absolutePath, config.inner.pemPrivateKeyPath);
        assertEquals("password", config.inner.privateKeyPassword);
    }

    @Test
    void loadPemFromPathWithPasswordTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromPath(absolutePathString,absolutePathString, "password");
        assertThrows(SSLConfigException.class,() -> config.pemFromPath(absolutePathString,absolutePathString, "password"));
    }

    //Repeat the same tests with the method loadPemFromClasspath
    @Test
    void loadPemFromClasspathCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromClasspath(fileName,fileName));
        assertEquals(fileName, config.inner.pemCertificatesFile);
        assertEquals(fileName, config.inner.pemPrivateKeyFile);
    }

    @Test
    void loadPemFromClasspathDifferentMethod() {
        SSLConfig config = new SSLConfig();
        config.pemFromString("", ""); // load empty strings
        assertThrows(SSLConfigException.class,() -> config.pemFromClasspath(fileName,fileName));
    }

    @Test
    void loadPemFromClasspathTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromClasspath(fileName,fileName);
        assertThrows(SSLConfigException.class,() -> config.pemFromClasspath(fileName,fileName));
    }

    @Test
    void loadPemFromClasspathWithPasswordCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromClasspath(fileName,fileName, "password"));
        assertEquals(fileName, config.inner.pemCertificatesFile);
        assertEquals(fileName, config.inner.pemPrivateKeyFile);
        assertEquals("password", config.inner.privateKeyPassword);
    }

    @Test
    void loadPemFromClasspathWithPasswordTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromClasspath(fileName,fileName, "password");
        assertThrows(SSLConfigException.class,() -> config.pemFromClasspath(fileName,fileName, "password"));
    }

    @Test
    void loadPemFromStringCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromString(pemString,pemString));
        assertEquals(pemString, config.inner.pemCertificatesString);
        assertEquals(pemString, config.inner.pemPrivateKeyString);
    }

    @Test
    void loadPemFromStringDifferentMethod() {
        SSLConfig config = new SSLConfig();
        config.pemFromPath(absolutePathString,absolutePathString); // load empty strings
        assertThrows(SSLConfigException.class,() -> config.pemFromString(pemString,pemString));
    }

    @Test
    void loadPemFromStringTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromString(pemString,pemString);
        assertThrows(SSLConfigException.class,() -> config.pemFromString(pemString,pemString));
    }

    @Test
    void loadPemFromStringWithPasswordCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromString(pemString,pemString, "password"));
        assertEquals(pemString, config.inner.pemCertificatesString);
        assertEquals(pemString, config.inner.pemPrivateKeyString);
        assertEquals("password", config.inner.privateKeyPassword);
    }

    @Test
    void loadPemFromStringWithPasswordTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromString(pemString,pemString, "password");
        assertThrows(SSLConfigException.class,() -> config.pemFromString(pemString,pemString, "password"));
    }


    //Repeat the same tests with the method loadPemFromInputStream
    @Test
    void loadPemFromInputStreamCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromInputStream(inputStream,inputStream));
        assertNotNull(config.inner.pemCertificatesInputStream);
        assertNotNull(config.inner.pemPrivateKeyInputStream);
    }

    @Test
    void loadPemFromInputStreamDifferentMethod() {
        SSLConfig config = new SSLConfig();
        config.pemFromString("", ""); // load empty strings
        assertThrows(SSLConfigException.class,() -> config.pemFromInputStream(inputStream,inputStream));
    }

    @Test
    void loadPemFromInputStreamTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromInputStream(inputStream,inputStream);
        assertThrows(SSLConfigException.class,() -> config.pemFromInputStream(inputStream,inputStream));
    }

    @Test
    void loadPemFromInputStreamWithPasswordCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.pemFromInputStream(inputStream,inputStream, "password"));
        assertNotNull(config.inner.pemCertificatesInputStream);
        assertNotNull(config.inner.pemPrivateKeyInputStream);
        assertEquals("password", config.inner.privateKeyPassword);
    }

    @Test
    void loadPemFromInputStreamWithPasswordTwice() {
        SSLConfig config = new SSLConfig();
        config.pemFromInputStream(inputStream,inputStream, "password");
        assertThrows(SSLConfigException.class,() -> config.pemFromInputStream(inputStream,inputStream, "password"));
    }


    //////////////////////////////
    // Keystore loading tests   //
    //////////////////////////////

    //Keystore loading tests are the same as the PEM loading tests, but with the keystore method instead of the pem method and always with a password
    @Test
    void loadKeystoreFromPathCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.keystoreFromPath(absolutePathString,"password"));
        assertEquals(absolutePath, config.inner.keyStorePath);
        assertEquals("password", config.inner.keyStorePassword);
    }

    @Test
    void loadKeystoreFromPathTwice() {
        SSLConfig config = new SSLConfig();
        config.keystoreFromPath(absolutePathString,"password");
        assertThrows(SSLConfigException.class,() -> config.keystoreFromPath(absolutePathString,"password"));
    }

    @Test
    void loadKeystoreFromInputStreamCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.keystoreFromInputStream(inputStream,"password"));
        assertNotNull(config.inner.keyStoreInputStream);
        assertEquals("password", config.inner.keyStorePassword);
    }

    @Test
    void loadKeystoreFromInputStreamTwice() {
        SSLConfig config = new SSLConfig();
        config.keystoreFromInputStream(inputStream,"password");
        assertThrows(SSLConfigException.class,() -> config.keystoreFromInputStream(inputStream,"password"));
    }

    @Test
    void loadKeystoreFromClasspathCorrectly() {
        SSLConfig config = new SSLConfig();
        assertDoesNotThrow(() -> config.keystoreFromClasspath(fileName,"password"));
        assertEquals(fileName, config.inner.keyStoreFile);
        assertEquals("password", config.inner.keyStorePassword);
    }

    @Test
    void loadKeystoreFromClasspathTwice() {
        SSLConfig config = new SSLConfig();
        config.keystoreFromClasspath(fileName,"password");
        assertThrows(SSLConfigException.class,() -> config.keystoreFromClasspath(fileName,"password"));
    }

}

File: src/test/kotlin/io/javalin/community/ssl/SSLConfigTest.kt
package io.javalin.community.ssl

import io.javalin.community.ssl.SSLConfigException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import java.io.InputStream
import java.nio.file.Paths

@Tag("unitary")
internal class SSLConfigTest {
    val absolutePathString = "/etc/sample/path"
    val absolutePath = Paths.get(absolutePathString)
    val fileName = "sample.pem"
    val inputStream = InputStream.nullInputStream()
    val pemString = """
        ----- START CERTIFICATE -----
        BLABLABLABLABLABLABLABLABLABLA----- END CERTIFICATE -----

        """.trimIndent()
    //Replace test names with a description of what the test does, using the fun ``() syntax
    //////////////////////////////
    // Pem loading tests        //
    //////////////////////////////
    @Test
    fun `load PEM from path correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromPath(absolutePathString, absolutePathString) }
        Assertions.assertEquals(absolutePath, config.inner.pemCertificatesPath)
        Assertions.assertEquals(absolutePath, config.inner.pemPrivateKeyPath)
    }

    @Test
    fun `load PEM from path with different method`() {
        val config = SSLConfig()
        config.pemFromString("", "") // load empty strings
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.pemFromPath(
                absolutePathString,
                absolutePathString
            )
        }
    }

    @Test
    fun `load PEM from path twice`() {
        val config = SSLConfig()
        config.pemFromPath(absolutePathString, absolutePathString)
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.pemFromPath(
                absolutePathString,
                absolutePathString
            )
        }
    }

    @Test
    fun `load PEM from path with password correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromPath(absolutePathString, absolutePathString, "password") }
        Assertions.assertEquals(absolutePath, config.inner.pemCertificatesPath)
        Assertions.assertEquals(absolutePath, config.inner.pemPrivateKeyPath)
        Assertions.assertEquals("password", config.inner.privateKeyPassword)
    }

    @Test
    fun `load PEM from path with password twice`() {
        val config = SSLConfig()
        config.pemFromPath(absolutePathString, absolutePathString, "password")
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.pemFromPath(
                absolutePathString,
                absolutePathString,
                "password"
            )
        }
    }

    //Repeat the same tests with the method loadPemFromClasspath
    @Test
    fun `load PEM from classpath correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromClasspath(fileName, fileName) }
        Assertions.assertEquals(fileName, config.inner.pemCertificatesFile)
        Assertions.assertEquals(fileName, config.inner.pemPrivateKeyFile)
    }

    @Test
    fun `load PEM from classpath with different method`() {
        val config = SSLConfig()
        config.pemFromString("", "") // load empty strings
        Assertions.assertThrows(SSLConfigException::class.java) { config.pemFromClasspath(fileName, fileName) }
    }

    @Test
    fun `load PEM from classpath twice`() {
        val config = SSLConfig()
        config.pemFromClasspath(fileName, fileName)
        Assertions.assertThrows(SSLConfigException::class.java) { config.pemFromClasspath(fileName, fileName) }
    }

    @Test
    fun `load PEM from classpath with password correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromClasspath(fileName, fileName, "password") }
        Assertions.assertEquals(fileName, config.inner.pemCertificatesFile)
        Assertions.assertEquals(fileName, config.inner.pemPrivateKeyFile)
        Assertions.assertEquals("password", config.inner.privateKeyPassword)
    }

    @Test
    fun `load PEM from classpath with password twice`() {
        val config = SSLConfig()
        config.pemFromClasspath(fileName, fileName, "password")
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.pemFromClasspath(
                fileName,
                fileName,
                "password"
            )
        }
    }

    @Test
    fun `load PEM from string correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromString(pemString, pemString) }
        Assertions.assertEquals(pemString, config.inner.pemCertificatesString)
        Assertions.assertEquals(pemString, config.inner.pemPrivateKeyString)
    }

    @Test
    fun `load PEM from string with different method`() {
        val config = SSLConfig()
        config.pemFromPath(absolutePathString, absolutePathString) // load empty strings
        Assertions.assertThrows(SSLConfigException::class.java) { config.pemFromString(pemString, pemString) }
    }

    @Test
    fun `load PEM from string twice`() {
        val config = SSLConfig()
        config.pemFromString(pemString, pemString)
        Assertions.assertThrows(SSLConfigException::class.java) { config.pemFromString(pemString, pemString) }
    }

    @Test
    fun `load PEM from string with password correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromString(pemString, pemString, "password") }
        Assertions.assertEquals(pemString, config.inner.pemCertificatesString)
        Assertions.assertEquals(pemString, config.inner.pemPrivateKeyString)
        Assertions.assertEquals("password", config.inner.privateKeyPassword)
    }

    @Test
    fun `load PEM from string with password twice`() {
        val config = SSLConfig()
        config.pemFromString(pemString, pemString, "password")
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.pemFromString(
                pemString,
                pemString,
                "password"
            )
        }
    }

    //Repeat the same tests with the method loadPemFromInputStream
    @Test
    fun `load PEM from input stream correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromInputStream(inputStream, inputStream) }
        Assertions.assertNotNull(config.inner.pemCertificatesInputStream)
        Assertions.assertNotNull(config.inner.pemPrivateKeyInputStream)
    }

    @Test
    fun `load PEM from input stream with different method`() {
        val config = SSLConfig()
        config.pemFromString("", "") // load empty strings
        Assertions.assertThrows(SSLConfigException::class.java) { config.pemFromInputStream(inputStream, inputStream) }
    }

    @Test
    fun `load PEM from input stream twice`() {
        val config = SSLConfig()
        config.pemFromInputStream(inputStream, inputStream)
        Assertions.assertThrows(SSLConfigException::class.java) { config.pemFromInputStream(inputStream, inputStream) }
    }

    @Test
    fun `load PEM from input stream with password correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.pemFromInputStream(inputStream, inputStream, "password") }
        Assertions.assertNotNull(config.inner.pemCertificatesInputStream)
        Assertions.assertNotNull(config.inner.pemPrivateKeyInputStream)
        Assertions.assertEquals("password", config.inner.privateKeyPassword)
    }

    @Test
    fun `load PEM from input stream with password twice`() {
        val config = SSLConfig()
        config.pemFromInputStream(inputStream, inputStream, "password")
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.pemFromInputStream(
                inputStream,
                inputStream,
                "password"
            )
        }
    }

    //////////////////////////////
    // Keystore loading tests   //
    //////////////////////////////
    @Test
    fun `load keystore from path correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.keystoreFromPath(absolutePathString, "password") }
        Assertions.assertEquals(absolutePath, config.inner.keyStorePath)
        Assertions.assertEquals("password", config.inner.keyStorePassword)
    }

    @Test
    fun `load keystore from path twice`() {
        val config = SSLConfig()
        config.keystoreFromPath(absolutePathString, "password")
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.keystoreFromPath(
                absolutePathString,
                "password"
            )
        }
    }

    @Test
    fun `load keystore from input stream correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.keystoreFromInputStream(inputStream, "password") }
        Assertions.assertNotNull(config.inner.keyStoreInputStream)
        Assertions.assertEquals("password", config.inner.keyStorePassword)
    }

    @Test
    fun `load keystore from input stream twice`() {
        val config = SSLConfig()
        config.keystoreFromInputStream(inputStream, "password")
        Assertions.assertThrows(SSLConfigException::class.java) {
            config.keystoreFromInputStream(
                inputStream,
                "password"
            )
        }
    }

    @Test
    fun `load keystore from classpath correctly`() {
        val config = SSLConfig()
        Assertions.assertDoesNotThrow { config.keystoreFromClasspath(fileName, "password") }
        Assertions.assertEquals(fileName, config.inner.keyStoreFile)
        Assertions.assertEquals("password", config.inner.keyStorePassword)
    }

    @Test
    fun `load keystore from classpath twice`() {
        val config = SSLConfig()
        config.keystoreFromClasspath(fileName, "password")
        Assertions.assertThrows(SSLConfigException::class.java) { config.keystoreFromClasspath(fileName, "password") }
    }
}
